plugins/promela/hu.bme.mit.gamma.xsts.promela.transformation/src/hu/bme/mit/gamma/xsts/promela/transformation/serializer/ModelSerializer.xtend

```diff
@@ -146,4 +177,97 @@ class ModelSerializer {
 			«subaction.serialize»
 		«ENDFOR»
 	'''
+	
+	def dispatch String serialize(ParallelAction action) {
+		val actionSize = action.actions.size
+		val actions = action.actions
+		
+		if (actionSize > 1) {
+			return '''
+				«FOR index : 0 ..< actionSize»
+					run Parallel_«parallelMapping.get(actions)»_«index»(«actions.get(index).serializeParallelProcCallArguments»);
+				«ENDFOR»
+				
+				«FOR index : 0 ..< actionSize»
+					chan_parallel_«index»?msg_parallel_«index»;
+					msg_parallel_«index» == 1;
+					msg_parallel_«index» = 0;
+				«ENDFOR»
+			'''
+		}
+		else {
+			return '''
+				«actions.get(0).serialize»
+			'''
+		} 
+	}
+	
+	def serializeParallelProcesses() '''
```

Comment:
All these parallel action handling methods should be refactored into a separate class